From: Dimitri John Ledkov <xnox@ubuntu.com>
Date: Wed, 22 May 2019 19:57:29 +0100
Subject: Add devicetree command, if a dtb is present.

Specically support dtb paths as installed by flash-kernel.

Signed-off-by: Dimitri John Ledkov <xnox@ubuntu.com>
Bug-Debian: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=929427
Bug-Upstream: https://lists.gnu.org/archive/html/grub-devel/2019-05/msg00121.html
Patch-Name: ubuntu-add-devicetree-command-support.patch
---
 util/grub.d/10_linux.in | 19 +++++++++++++++++++
 1 file changed, 19 insertions(+)

diff --git a/util/grub.d/10_linux.in b/util/grub.d/10_linux.in
index ac6226f..2dc026b 100644
--- a/util/grub.d/10_linux.in
+++ b/util/grub.d/10_linux.in
@@ -254,6 +254,17 @@ EOF
 EOF
         fi
     fi
+    if test -n "${dtb}" ; then
+      if [ x"$quiet_boot" = x0 ] || [ x"$type" != xsimple ]; then
+        message="$(gettext_printf "Loading device tree blob...")"
+        sed "s/^/$submenu_indentation/" << EOF
+	echo	'$(echo "$message" | grub_quote)'
+EOF
+      fi
+      sed "s/^/$submenu_indentation/" << EOF
+	devicetree	${rel_dirname}/${dtb}
+EOF
+    fi
   fi
   sed "s/^/$submenu_indentation/" << EOF
 }
@@ -389,6 +400,14 @@ while [ "x$list" != "x" ] ; do
     gettext_printf "Found initrd image: %s\n" "$(echo $initrd_display)" >&2
   fi
 
+  dtb=
+  for i in "dtb-${version}" "dtb-${alt_version}" "dtb"; do
+    if test -e "${dirname}/${i}" ; then
+      dtb="$i"
+      break
+    fi
+  done
+
   config=
   for i in "${dirname}/config-${version}" "${dirname}/config-${alt_version}" "/etc/kernels/kernel-config-${version}" ; do
     if test -e "${i}" ; then
